src/mscorlib/src/System/Globalization/CompareInfo.Unix.cs

```diff
@@ -284,7 +313,136 @@ private bool EndsWith(string source, string suffix, CompareOptions options)
 
             return Interop.Globalization.EndsWith(_sortHandle, suffix, suffix.Length, source, source.Length, options);
         }
-        
+
+        private unsafe bool EndsWith(ReadOnlySpan<char> source, ReadOnlySpan<char> suffix, CompareOptions options)
+        {
+            Debug.Assert(!_invariantMode);
+
+            Debug.Assert(!source.IsEmpty);
+            Debug.Assert(!suffix.IsEmpty);
+            Debug.Assert((options & (CompareOptions.Ordinal | CompareOptions.OrdinalIgnoreCase)) == 0);
+
+            if (_isAsciiEqualityOrdinal && CanUseAsciiOrdinalForOptions(options))
+            {
+                if ((options & CompareOptions.IgnoreCase) == CompareOptions.IgnoreCase)
+                {
+                    return EndsWithOrdinalIgnoreCaseHelper(source, suffix, options);
+                }
+                else
+                {
+                    return EndsWithOrdinalHelper(source, suffix, options);
+                }
+            }
+            else
+            {
+                fixed (char* pSource = &MemoryMarshal.GetReference(source))
+                fixed (char* pSuffix = &MemoryMarshal.GetReference(suffix))
+                {
+                    return Interop.Globalization.EndsWith(_sortHandle, pSuffix, suffix.Length, pSource, source.Length, options);
+                }
+            }
+        }
+
+        private unsafe bool EndsWithOrdinalIgnoreCaseHelper(ReadOnlySpan<char> source, ReadOnlySpan<char> suffix, CompareOptions options)
+        {
+            return StartsWithOrdinalIgnoreCaseHelper(source.Slice(source.Length - suffix.Length), suffix, options);
+        }
+
+        private unsafe bool EndsWithOrdinalHelper(ReadOnlySpan<char> source, ReadOnlySpan<char> suffix, CompareOptions options)
+        {
+            return StartsWithOrdinalHelper(source.Slice(source.Length - suffix.Length), suffix, options);
+        }
+
+        private unsafe bool StartsWithOrdinalIgnoreCaseHelper(ReadOnlySpan<char> source, ReadOnlySpan<char> prefix, CompareOptions options)
+        {
+            Debug.Assert(!_invariantMode);
+
+            Debug.Assert(!source.IsEmpty);
+            Debug.Assert(!prefix.IsEmpty);
+            Debug.Assert(_isAsciiEqualityOrdinal);
+            Debug.Assert(source.Length >= prefix.Length);
+
+            int length = prefix.Length;
+
+            fixed (char* ap = &MemoryMarshal.GetReference(source))
+            fixed (char* bp = &MemoryMarshal.GetReference(prefix))
+            {
+                char* a = ap;
+                char* b = bp;
+
+                while (length != 0 && (*a <= 0x80) && (*b <= 0x80))
```

Comment:
Should this rather be `< 0x80` ?

Reply:
Also, the string implementations use IsFastSort. Fast sort is false even for certain characters `< 0x80`.

Reply:
That is what I though, but I borrowed the implementation from here which also has <= 0x80:
https://github.com/dotnet/coreclr/blob/master/src/mscorlib/shared/System/Globalization/CompareInfo.cs#L569

```C#
                char maxChar = (char) (GlobalizationMode.Invariant ? 0xFFFF : 0x80);

                while (length != 0 && (*a <= maxChar) && (*b <= maxChar))
```

Reply:
> Fast sort is false even for certain characters

https://github.com/dotnet/coreclr/blob/ae3b92ad2ade04709abb6d4846018d0a26b872c8/src/vm/object.cpp#L1369

> The requirements are that ... the string not contain an apostrophe or a hypen.  Apostrophe and hyphen are excluded so that words like co-op and coop sort together.

This affects sorting (and hence CompareTo/etc). But I don't think this should affect equality operations like StartsWith/EndsWith.

Reply:
After discussing with @tarekgh, I realized my misunderstanding. I will update the implementation.